File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/stories/StoryItemMenuSheet.kt
/*
 * StoryItemMenuSheet.kt
 *
 * Per-item ⋯ menu from §7.3b. Mirror of iOS `StoryItemMenuSheet.swift`.
 */

@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package app.quickink.mobile.features.stories

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowDownward
import androidx.compose.material.icons.filled.ArrowUpward
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.outlined.Delete
import androidx.compose.material.icons.outlined.Edit
import androidx.compose.material.icons.outlined.GridView
import androidx.compose.material.icons.outlined.StarBorder
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.quickink.mobile.data.storyitem.StoryItemEntity
import app.quickink.mobile.ui.theme.LocalQuickInkColors
import app.quickink.mobile.ui.theme.LocalQuickInkTypography
import app.quickink.mobile.ui.theme.QuickInkRadius
import app.quickink.mobile.ui.theme.QuickInkSpacing
import kotlinx.coroutines.delay

@Composable
fun StoryItemMenuSheet(
    item: StoryItemEntity,
    isCoverItem: Boolean,
    onSetAsCover: () -> Unit,
    onLayoutChange: (StoryItemEntity.Layout) -> Unit,
    onMoveUp: () -> Unit,
    onMoveDown: () -> Unit,
    onRemove: () -> Unit,
    onDismiss: () -> Unit,
) {
    val colors = LocalQuickInkColors.current
    val type   = LocalQuickInkTypography.current
    val sheetState = rememberModalBottomSheetState()
    var stubToast by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(stubToast) {
        if (stubToast != null) {
            delay(1_800)
            stubToast = null
        }
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState       = sheetState,
        containerColor   = colors.surface,
    ) {
        Column(modifier = Modifier.padding(bottom = QuickInkSpacing.s4)) {
            // Item preview
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = QuickInkSpacing.s4, vertical = QuickInkSpacing.s2)
                    .border(0.5.dp, colors.borderSoft, RoundedCornerShape(0.dp)),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(QuickInkSpacing.s3),
            ) {
                Box(
                    modifier = Modifier
                        .size(48.dp)
                        .clip(RoundedCornerShape(8.dp))
                        .background(colors.paper1),
                )
                Column(modifier = Modifier.weight(1f)) {
                    Text(previewTitle(item), style = type.editorial, color = colors.ink, maxLines = 1)
                    Text(previewSubtitle(item), style = type.bodyItalic, color = colors.inkSoft, maxLines = 1)
                }
            }
            Box(modifier = Modifier.fillMaxWidth().height(0.5.dp).background(colors.borderSoft))

            ActionRow(icon = Icons.Outlined.Edit, label = "Edit caption",
                onClick = { stubToast = "Tap a caption inline to edit" })
            ActionRow(
                icon = if (isCoverItem) Icons.Filled.Star else Icons.Outlined.StarBorder,
                label = if (isCoverItem) "Cover item" else "Set as cover",
                onClick = onSetAsCover,
            )

            // Layout pills
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = QuickInkSpacing.s4, vertical = QuickInkSpacing.s2),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(Icons.Outlined.GridView, contentDescription = null, tint = colors.inkSoft, modifier = Modifier.size(20.dp))
                Spacer(modifier = Modifier.width(QuickInkSpacing.s2))
                Text("Layout", style = type.body, color = colors.ink)
                Spacer(modifier = Modifier.weight(1f))
                Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                    LayoutPill(label = "Full",  selected = item.layout == StoryItemEntity.Layout.FULL.raw,  onClick = { onLayoutChange(StoryItemEntity.Layout.FULL) })
                    LayoutPill(label = "Half",  selected = item.layout == StoryItemEntity.Layout.HALF.raw,  onClick = { onLayoutChange(StoryItemEntity.Layout.HALF) })
                    LayoutPill(label = "Grid",  selected = item.layout == StoryItemEntity.Layout.GRID.raw,  onClick = { onLayoutChange(StoryItemEntity.Layout.GRID) })
                }
            }

            Divider()
            ActionRow(icon = Icons.Filled.Refresh, label = "Replace with another item",
                onClick = { stubToast = "Replace coming soon" })
            ActionRow(icon = Icons.Filled.ArrowUpward,   label = "Move up",   onClick = onMoveUp)
            ActionRow(icon = Icons.Filled.ArrowDownward, label = "Move down", onClick = onMoveDown)
            Divider()
            ActionRow(icon = Icons.Outlined.Delete, label = "Remove from story",
                tint = colors.accent, onClick = onRemove)

            if (stubToast != null) {
                Text(stubToast!!, style = type.bodyItalic, color = colors.inkSoft,
                    modifier = Modifier.padding(horizontal = QuickInkSpacing.s4, vertical = QuickInkSpacing.s2))
            }
        }
    }
}

@Composable
private fun ActionRow(
    icon: ImageVector,
    label: String,
    tint: Color? = null,
    onClick: () -> Unit,
) {
    val colors = LocalQuickInkColors.current
    val type   = LocalQuickInkTypography.current
    val effectiveTint = tint ?: colors.ink
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(horizontal = QuickInkSpacing.s4, vertical = QuickInkSpacing.s2 + 2.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(imageVector = icon, contentDescription = null, tint = effectiveTint, modifier = Modifier.size(20.dp).padding(end = 0.dp))
        Spacer(modifier = Modifier.width(QuickInkSpacing.s2))
        Text(label, style = type.body, color = effectiveTint)
    }
}

@Composable
private fun LayoutPill(label: String, selected: Boolean, onClick: () -> Unit) {
    val colors = LocalQuickInkColors.current
    val bg = if (selected) colors.accentSoft else Color.Transparent
    val border = if (selected) colors.accent.copy(alpha = 0.4f) else colors.border
    val fg = if (selected) colors.accent else colors.inkSoft
    Box(
        modifier = Modifier
            .clip(RoundedCornerShape(QuickInkRadius.pill))
            .background(bg)
            .border(1.dp, border, RoundedCornerShape(QuickInkRadius.pill))
            .clickable(onClick = onClick)
            .padding(horizontal = 10.dp, vertical = 5.dp),
    ) {
        Text(label, color = fg, fontSize = 11.sp, fontWeight = FontWeight.Medium)
    }
}

@Composable
private fun Divider() {
    val colors = LocalQuickInkColors.current
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(0.5.dp)
            .background(colors.borderSoft)
            .padding(vertical = QuickInkSpacing.s1, horizontal = QuickInkSpacing.s4),
    )
}

private fun previewTitle(item: StoryItemEntity): String = when (item.kind) {
    StoryItemEntity.Kind.TEXT_BLOCK.raw        -> item.text ?: "Paragraph"
    StoryItemEntity.Kind.HANDWRITTEN_NOTE.raw  -> item.text ?: "Handwritten note"
    StoryItemEntity.Kind.DATE_DIVIDER.raw      -> item.text ?: "Date divider"
    StoryItemEntity.Kind.PLACE_PIN.raw         -> item.text ?: "Place pin"
    StoryItemEntity.Kind.VOICE_CLIP.raw        -> "Voice clip"
    else                                       -> item.caption ?: "Item"
}

private fun previewSubtitle(item: StoryItemEntity): String =
    item.caption ?: item.kind.replace('_', ' ')

File: apps/quickink/ios/QuickInk/Stories/StoryItemMenuSheet.swift
/*
 * StoryItemMenuSheet.swift
 *
 * Per-item ⋯ menu from §7.3b of the v3 mockup. Item preview at the
 * top anchors which item is being acted on; Edit caption / Set as
 * cover rows; inline Layout pills (Full / Half / Grid); Replace + reorder
 * (Replace stub for Phase 2 follow-up, Move up/down via repo's
 * `commitReorder`); destructive Remove below a divider.
 *
 * Mirror of Android `StoryItemMenuSheet.kt`.
 */

import SwiftUI

struct StoryItemMenuSheet: View {

    let item: StoryItem
    let isCoverItem: Bool
    var onEditCaption: () -> Void
    var onSetAsCover: () -> Void
    var onLayoutChange: (StoryItem.Layout) -> Void
    var onMoveUp: () -> Void
    var onMoveDown: () -> Void
    var onRemove: () -> Void

    @State private var stubToast: String? = nil

    var body: some View {
        VStack(spacing: 0) {
            handle
            preview
            VStack(spacing: 0) {
                actionRow(icon: "pencil",  label: "Edit caption",     action: onEditCaption)
                actionRow(
                    icon: isCoverItem ? "star.fill" : "star",
                    label: isCoverItem ? "Cover item" : "Set as cover",
                    action: onSetAsCover
                )

                HStack(spacing: QuickInkSpacing.s2) {
                    Image(systemName: "rectangle.grid.2x2")
                        .foregroundStyle(QuickInkColors.inkSoft)
                        .frame(width: 24)
                    Text("Layout")
                        .font(QuickInkText.body)
                        .foregroundStyle(QuickInkColors.ink)
                    Spacer()
                    layoutPills
                }
                .padding(.horizontal, QuickInkSpacing.s4)
                .padding(.vertical, QuickInkSpacing.s2)

                divider
                actionRow(icon: "arrow.triangle.2.circlepath", label: "Replace with another item",
                          action: stub("Replace coming soon"))
                actionRow(icon: "arrow.up", label: "Move up",   action: onMoveUp)
                actionRow(icon: "arrow.down", label: "Move down", action: onMoveDown)

                divider
                actionRow(icon: "trash", label: "Remove from story",
                          tint: QuickInkColors.accent, action: onRemove)
            }
            if let toast = stubToast {
                Text(toast)
                    .font(QuickInkText.bodyItalic)
                    .foregroundStyle(QuickInkColors.inkSoft)
                    .padding(.bottom, QuickInkSpacing.s3)
            }
        }
        .padding(.bottom, QuickInkSpacing.s4)
        .background(QuickInkColors.surface)
    }

    // MARK: - Pieces

    private var handle: some View {
        Capsule()
            .fill(QuickInkColors.border)
            .frame(width: 38, height: 4)
            .padding(.top, QuickInkSpacing.s2)
            .padding(.bottom, QuickInkSpacing.s3)
    }

    private var preview: some View {
        HStack(spacing: QuickInkSpacing.s3) {
            RoundedRectangle(cornerRadius: 8)
                .fill(QuickInkColors.paper1)
                .frame(width: 48, height: 48)
            VStack(alignment: .leading, spacing: 2) {
                Text(previewTitle)
                    .font(QuickInkText.editorial)
                    .foregroundStyle(QuickInkColors.ink)
                    .lineLimit(1)
                Text(previewSubtitle)
                    .font(QuickInkText.bodyItalic)
                    .foregroundStyle(QuickInkColors.inkSoft)
                    .lineLimit(1)
            }
            Spacer()
        }
        .padding(.horizontal, QuickInkSpacing.s4)
        .padding(.vertical, QuickInkSpacing.s2)
        .overlay(alignment: .bottom) {
            Rectangle().fill(QuickInkColors.borderSoft).frame(height: 0.5)
        }
    }

    private var layoutPills: some View {
        HStack(spacing: 4) {
            ForEach([StoryItem.Layout.full, .half, .grid], id: \.self) { layout in
                Button(action: { onLayoutChange(layout) }) {
                    Text(layout.rawValue.capitalized)
                        .font(.system(size: 11, weight: .medium))
                        .padding(.horizontal, 10)
                        .padding(.vertical, 5)
                        .background(
                            Capsule().fill(
                                item.layout == layout.rawValue
                                    ? QuickInkColors.accentSoft
                                    : Color.clear
                            )
                        )
                        .overlay(
                            Capsule().strokeBorder(
                                item.layout == layout.rawValue
                                    ? QuickInkColors.accent.opacity(0.4)
                                    : QuickInkColors.border,
                                lineWidth: 1
                            )
                        )
                        .foregroundStyle(
                            item.layout == layout.rawValue
                                ? QuickInkColors.accent
                                : QuickInkColors.inkSoft
                        )
                }
                .buttonStyle(.plain)
            }
        }
    }

    private func actionRow(
        icon: String,
        label: String,
        tint: Color = QuickInkColors.ink,
        action: @escaping () -> Void
    ) -> some View {
        Button(action: action) {
            HStack(spacing: QuickInkSpacing.s2) {
                Image(systemName: icon)
                    .foregroundStyle(tint)
                    .frame(width: 24)
                Text(label)
                    .font(QuickInkText.body)
                    .foregroundStyle(tint)
                Spacer()
            }
            .padding(.horizontal, QuickInkSpacing.s4)
            .padding(.vertical, QuickInkSpacing.s2 + 2)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    private var divider: some View {
        Rectangle()
            .fill(QuickInkColors.borderSoft)
            .frame(height: 0.5)
            .padding(.vertical, QuickInkSpacing.s1)
            .padding(.horizontal, QuickInkSpacing.s4)
    }

    private var previewTitle: String {
        switch item.kind {
        case StoryItem.Kind.textBlock.rawValue:        return item.text ?? "Paragraph"
        case StoryItem.Kind.handwrittenNote.rawValue:  return item.text ?? "Handwritten note"
        case StoryItem.Kind.dateDivider.rawValue:      return item.text ?? "Date divider"
        case StoryItem.Kind.placePin.rawValue:         return item.text ?? "Place pin"
        case StoryItem.Kind.voiceClip.rawValue:        return "Voice clip"
        default:                                       return item.caption ?? "Item"
        }
    }

    private var previewSubtitle: String {
        item.caption ?? item.kind.replacingOccurrences(of: "_", with: " ")
    }

    private func stub(_ message: String) -> () -> Void {
        return {
            stubToast = message
            Task { @MainActor in
                try? await Task.sleep(nanoseconds: 1_800_000_000)
                stubToast = nil
            }
        }
    }
}
